keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2262 @@
+
+A quota system that limits how many resources a user is allowed to allocate
+needs to be supported by the resource driver. When a user has exhausted their
+quota, the driver then would refuse to allocate further ResourceClaims. Such a
+quota system cannot be implemented in core Kubernetes because Kubernetes has no
+information about how much a certain ResourceClaim count against the quota.
+
+#### Usability
+
+Aside from security implications, usability and usefulness of dynamic resource
+allocation also may turn out to be insufficient. Some risks are:
+
+- Slower pod scheduling due to the interaction with resource drivers.
+
+- Additional complexity when describing pod requirements because
+  separate objects must be created for the parameters.
+
+- Network-attached resources may have additional constraints that are not
+  captured yet (like limited number of nodes that they can be attached to).
+
+- Cluster autoscaling will not work as expected unless the autoscaler and
+  resource drivers get extended to support it.
+
+All of these risks will have to be evaluated by gathering feedback from users
+and resource driver developers.
+
+## Design Details
+
+### Components
+
+![components](./components.png)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which creates 
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral.
+  It also removes the reservation entry for a user in a ResourceClaim (`ReservedFor` field)
+  when that user no longer exists.
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and to call a resource kubelet plugin. That plugin returns CDI device ID(s)
+  which then must be passed to the container runtime.
+
+For a resource driver the following components are needed:
+- *Resource driver controller*: a central component which handles resource allocation
+  by watching ResourceClaims and updating their status once it is done with
+  allocation. It may run inside the cluster or outside of it. The only
+  hard requirement is that it can connect to the API server. Optionally,
+  it may also be configured as a [special scheduler extender](#kube-scheduler).
+- *Resource kubelet plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+An utility library for resource drivers will be developed outside of Kubernetes
+and does not have to be used by drivers, therefore it is not described further
+in this KEP.
+
+### State and communication
```

Comment:
Before you fall into all these details, I really feel like a "theory of operation" section is missing.  E.g.

In general, this new API works as described below (more details and exploration of corner cases will follow):

* A user creates a ResourceClaim for a specific ResourceClass.  This claim may be created by the user directly (the user owns the resource's lifetime) or indirectly through a pod (the pod owns the resource's lifetime).
* A resource driver observes the claim and allocates the underlying resource if it can
* In the case of late-binding (the pod owns the lifetime), the kube scheduler and the resource driver(s) coordinate to pick a viable node
* Once allocated, a pod which consumes the resource is scheduled to a node
* Kubelet on that node communicates with a node-resource driver to present the resource to the pod
* When the pod completes, the Kubelet on that node communicates with a node-resource driver to clean up
* When the claim is released (deleted), the resource driver can free the underlying resource.

THEN you can talk about params, immutability, etc.  I wrote the above after reading the KEP 10 times, and I am still not sure I got it all right.

Reply:
You did get it right and phrased it so well that I'll start with this and then will just update the wording a bit ("late-binding" is for volumes, etc.).

I put this intro even before "Components".